examples/layers/examples/simple_retry.rs

```diff
@@ -0,0 +1,154 @@
+        Self { max_retries: AtomicU64::new(max_retries), backoff_interval: self.backoff_interval }
+    }
+}
+
+impl Policy<RequestPacket, ResponsePacket, TransportError> for RetryPolicy {
+    type Future = Pin<Box<dyn Future<Output = Self> + Send + 'static>>;
+
+    fn retry(
+        &self,
+        req: &RequestPacket,
+        result: Result<&ResponsePacket, &TransportError>,
+    ) -> Option<Self::Future> {
+        // Retry on any error for testing.
+        // TODO: Use rate-limit specific errors/codes and retry accordingly.
+        if result.is_err() {
+            let max_retries = self.max_retries.load(Ordering::Relaxed);
+            if max_retries > 0 {
+                self.max_retries.store(max_retries - 1, Ordering::Relaxed);
+                if let RequestPacket::Single(req) = req {
+                    println!("retrying request {:?}", req.meta().method);
+                }
+                Some(Box::pin(future::ready(self.clone())))
+            } else {
+                println!(
+                    "max_retries exhausted, sleeping for: {:?}ms",
+                    self.backoff_interval.as_millis()
+                );
+                std::thread::sleep(self.backoff_interval);
+                Some(Box::pin(future::ready(self.clone())))
+            }
+        } else {
+            println!("success, no retry");
+            None
+        }
+    }
+
+    fn clone_request(&self, req: &RequestPacket) -> Option<RequestPacket> {
+        Some(req.clone())
+    }
+}
+
+/// RetryLayer
+struct RetryLayer {
+    policy: RetryPolicy,
+}
+
+impl RetryLayer {
+    const fn new(policy: RetryPolicy) -> Self {
+        Self { policy }
+    }
+}
+
+impl<S> Layer<S> for RetryLayer {
+    type Service = RetryService<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        RetryService { inner, policy: self.policy.clone() }
+    }
+}
+
+/// RetryService
+#[derive(Debug, Clone)]
+struct RetryService<S> {
+    inner: S,
+    policy: RetryPolicy,
+}
+impl<S> Service<RequestPacket> for RetryService<S>
+where
+    S: Service<RequestPacket, Response = ResponsePacket, Error = TransportError>
+        + Send
+        + 'static
+        + Clone,
+    S::Future: Send + 'static,
+{
+    type Response = ResponsePacket;
+    type Error = TransportError;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: RequestPacket) -> Self::Future {
+        let inner = self.inner.clone();
+        let policy = self.policy.clone();
+
+        let mut inner = std::mem::replace(&mut self.inner, inner);
+        Box::pin(async move {
+            let mut retries = 0;
+
+            let mut res = inner.call(req.clone()).await;
+
+            while let Some(_policy) = policy.retry(&req, res.as_ref()) {
```

Comment:
I think the policy should be updated here such that any changes to the policy on an iteration take affect. See [here](https://github.com/fractal-zkp/zero-bin/blob/2cbcf98ea8bf57ee17775d2e72c51dd40e8cd04b/rpc/src/retry.rs#L119-L135) for an example.